Feature/59178 journalize changes to project stages and gates

toy opened the pull request:
# Ticket
[OP#59178](https://community.openproject.org/wp/59178)

# What are you trying to accomplish?
* Journal activation/deactivation of project life cycle steps and setting/changing/removing their dates.
* Show those journals when «Project details» is selected in activities event type filter

# What approach did you choose and why?
* Not using feature flag, as both activation and editing of project life cycle steps are already hidden behind it
* Not creating initial journals, as project stages and gates should not yet be used by anyone
* Renamed event type itself from `project_attributes` to `project_details`, as project attributes is the name for project custom fields. Name `project_details` is also used in few migrations, but that is much less then usage of `project_attributes`. Created a compatibility alias for users that saved the link or have event types stored in session. [OP#61285](https://community.openproject.org/wp/61285) to cleanup.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

toy committed:
[#59178] Journalize changes to project stages and gates

https://community.openproject.org/work_packages/59178

toy committed:
correct memoization in successor and predecessor of Journal

toy committed:
rename activity filter from project attribute to project details

toy committed:
fix indent

toy committed:
sort associated journal relations

toy committed:
inline sql variables unless too many bindings

toy committed:
extract journable_id

toy committed:
sort supported associations

toy committed:
project life cycle step journal model and migration

toy committed:
add project life cycle steps to journal create service

toy committed:
remove empty factory blocks

toy committed:
simplify equality checks and frozen string literal in jounrable differ spec

toy committed:
spec for association_changes_multiple_attributes

toy committed:
remove overwritten assignment

toy committed:
keep order of original and changed same

toy committed:
simplify normalize_newlines using transform_values

toy committed:
refactor no_nil_to_empty_strings?

toy committed:
get old_value once

toy committed:
frozen string literal

toy committed:
inline changes_on_association

toy committed:
simplify filtering of changes

toy committed:
simplify and inline to_changes_format

toy committed:
extract create service associations code

toy committed:
extract model changes to module Differ::Model

toy committed:
rename variables for clarity

toy committed:
replace JournableDiffer leftovers with Differ::Association

toy committed:
allow getting changes grouped by association model

toy committed:
register adding project_life_cycle_steps_changes

toy committed:
test project changes to deactivated custom fields

toy committed:
filter for activated custom fields when querying customizable journals instead of afterwards

toy committed:
create journal when enabling/disabling life cycle steps

toy committed:
specify that absence of value for assocation diff returns empty string and not nil

toy committed:
get attributes using method calls instead of attributes hash to allow "virtual" attributes

toy committed:
combine association single attribute specs

toy committed:
options to get single value changes, or non joined changes of association

toy committed:
test multiple values for same id field

toy committed:
displaying project life cycle step change

toy committed:
moved journal_formatter specs under open_project

toy committed:
clearer differ association method names

toy committed:
document Differ::Association methods and rename all_values to multiple_values

toy committed:
stricter check (including order) for project journal changes

toy committed:
test porject life cycle step journaling

toy committed:
refactor project attributes activity feature spec

toy committed:
test activity when activating or changing dates of project life cycle steps

toy committed:
ensure that scope is an array, to not error if it is not

toy committed:
rename activity type project_attributes to project_details

toy committed:
erblint + use label_tag helper

toy committed:
activity provider event_params should still return a Hash

toy committed:
rename events.project_attributes to events.project_details

toy committed:
rework sorting of activity events

toy committed:
simplify understanding errors by comparing objects instead of their ids

toy committed:
cleanup before adding scope tests

toy committed:
test activities fetcher scopes

toy committed:
handle old project_attributes event type filter

toy committed:
simplify scope parsing

dombesz commented on app/models/activities/base_activity_provider.rb:
```diff
@@ -187,6 +187,7 @@ def event_params(event_data)
     params
   rescue StandardError => e
     Rails.logger.error "Failed to deduce event params for #{event_data.inspect}: #{e}"
+    {}
```
Was this a bug?

toy replied:
It was intentional, but I'm not sure if there is much sense in it. Normally the method returns a `Hash`, but in case of exception it logs it and returns `true` from `Rails.logger.error` which is unexpected by the only use in `fill_events`. So returning empty `Hash` prevents rescuing one exception to cause another. And I am not sure about it all, as this just hides the exception, and I don't know if there are maybe legacy cases for errors that should be hidden. So maybe the correct way is just to remove rescue and logging altogether.

toy replied:
Removed it

dombesz commented on app/models/activities/fetcher.rb:
```diff
@@ -72,22 +72,20 @@ def events(from: nil, to: nil, limit: nil)
       case scope
-      when :all
-        @scope = event_types
+      when :all, false, nil
```
Not sure about these values, should we also include other possible values, like `true`?

toy replied:
This handles the case when scope is falsey from `options[:scope] || :all`

dombesz commented on app/models/journal/project_life_cycle_step_journal.rb:
```diff
@@ -0,0 +1,41 @@
+  alias_attribute :date, :start_date
+
+  def date_range
+    start_date..end_date if start_date || end_date
```
Is this a typo, should it be `if start_date && end_date` ? This will generate open ended ranges.

toy replied:
This was intended, maybe not really warranted, but I'm trying to prevent hiding data if in some conditions only one of dates is assigned

dombesz commented on app/services/journals/create_service/association.rb:
```diff
@@ -0,0 +1,55 @@
+    include Helpers
+
+    def self.for(journable)
+      [
+        AgendaItemable,
+        Attachable,
+        Customizable,
+        ProjectLifeCycleStep,
+        Storable
+      ]
```
Should this be moved into a constant?

toy replied:
I agree, just rails dev reloading makes it wrong in any way: assigning classes to a constant causes constant to have outdated classes if they are reloaded, assigning names makes it ugly especially with namespaces, and keeping them in a method warrants a valid question why aren't they in a constant

dombesz commented on app/services/journals/create_service/helpers.rb:
```diff
@@ -0,0 +1,66 @@
+    delegate :sanitize,
+             to: ::OpenProject::SqlSanitization
+
+    def journable_id = journable.id
```
It seems quite arbitrary to move the `journable_id` into the helper and not move the `journable_type` method.
I found a few more methods that could also be part of the helper: `data_sink_columns`, `data_source_columns`, `journable_data_sql_addition`, `text_column_names`, `journable_timestamp`, `journable_type`, `journable_table_name`, `data_table_name`, `cause_sql`.

toy replied:
I extracted only methods that needed to be shared between classes handling associations sql and `CreateService`, but looking again it will be cleaner if delegation is used

dombesz commented on lib_static/plugins/acts_as_journalized/lib/acts/journalized/differ/association.rb:
```diff
@@ -0,0 +1,124 @@
+    # @param changed [ActiveRecord::Base] the changed object containing the association.
+    # @param association [Symbol] the name of the association to be compared.
+    # @param id_attribute [Symbol] the attribute used to correlate association models.
+    # @param multiple_values [Boolean] whether there are multiple values per id_attribute:
```
Just a small note, the type here should be `[Boolean, :joined]`

dombesz commented on lib_static/plugins/acts_as_journalized/lib/acts/journalized/differ/association.rb:
```diff
@@ -0,0 +1,124 @@
+            result["#{key_prefix}_#{id}"][attribute] = change
+          else
+            result["#{key_prefix}_#{id}_#{attribute}"] = change
+          end
```
I believe introducing 2 different key formats adds unnecessary complexity, I would rather prefer to have one key format only.

toy replied:
This was really necessary until I introduced diffing virtual attributes instead of attributes only, but now I think it should be possible to adapt the code to not need grouped diff

dombesz commented on lib/open_project/journal_formatter/project_life_cycle_step.rb:
```diff
@@ -0,0 +1,90 @@
+  def render(key, values, options = { html: true })
+    step = Project::LifeCycleStep.find(key[/\d+/])
+
+    name = step.definition.name
```
These will create 2 * M * N + 1 queries. I would suggest taking a look at the [JournalFormatterCache](https://github.com/opf/openproject/blob/dev/lib_static/plugins/acts_as_journalized/lib/journal_formatter_cache.rb) and perhaps at the [JournalFormatter::NamedAssociation](https://github.com/opf/openproject/blob/dev/lib_static/plugins/acts_as_journalized/lib/journal_formatter/named_association.rb) to eliminate it.

toy replied:
The problem is present for all formatters requiring to load additional models, so I've extracted solving it into https://community.openproject.org/wp/62330

dombesz commented on spec/features/projects/life_cycle/active_in_project_spec.rb:
```diff
@@ -56,10 +56,32 @@
 
   let(:project_life_cycle_page) { Pages::Projects::Settings::LifeCycle.new(project) }
 
+  let(:activity_page) { Pages::Projects::Activity.new(project) }
```
I'm not very firm on this one, but having the activity related specs under the `/features/activities/` would make it more consistent with the existing specs.

toy replied:
This is here only to check that activity is created after activating life cycles

dombesz commented on spec/features/projects/life_cycle/overview_page/dialog/update_spec.rb:
```diff
@@ -31,10 +31,13 @@
 
   let(:overview_page) { Pages::Projects::Show.new(project) }
 
+  let(:activity_page) { Pages::Projects::Activity.new(project) }
```
Same here, I would prefer the activity related specs under the `/features/activities/`.

toy replied:
And here only to check that activity is created after changing life cycle dates

dombesz commented on spec/lib/acts/journalized/differ/association_spec.rb:
```diff
@@ -0,0 +1,257 @@
+              "project_life_cycle_steps_1_active" => ["false", "true"],
+              "project_life_cycle_steps_2_active" => [nil, "true"],
+              "project_life_cycle_steps_3_active" => ["true", "false"],
+              "project_life_cycle_steps_3_end_date" => ["", "2024-01-18"],
+              "project_life_cycle_steps_3_start_date" => ["", "2024-01-17"],
```
Is there any reason the `end_date` comes first in all the expectations? It makes it a bit harder to read IMO.

toy replied:
Alphabetical :) But I agree that it is easier to read the other way

dombesz commented on spec/lib/acts/journalized/differ/association_spec.rb:
```diff
@@ -0,0 +1,257 @@
+      end
+
+      describe "requesting single value change" do
+        let(:multiple_values) { false }
```
The `let(:multiple_values) { true }` case does not seem to be covered.

toy replied:
I skipped that combination, as it is not used and supposedly doesn't need to be used.
The `:joined` is reproducing the behaviour before this PR which is used by all other formatters, where expectation is that all values are strings and multiple and joined with comma. So I added a way to not convert dates to string only to parse them back. But from what I saw multiple values are only needed when getting single attribute for custom fields, so I skipped it for `multiple_attributes_changes` tests.

dombesz reviewed (changes requested):
I found a few minor issues and left comments about them. Notably the N+1 issues would require attention, and the scenario below also produces interesting results:

1. Enable a stage
2. Set the date interval
3. Disable the stage

Result: The activities show only the date interval update from step 2. The activation and deactivation does not show up as they cancel each other out.
This looks misleading as the value can only be updated if the step is activated, and the logs show nothing about the field being active. One would assume the field was activated in a previous activity, but that is incorrect.

I propose to update the [`def aggregatable?`](https://github.com/opf/openproject/blob/21b775d1237ed5561341cdf981dab785e4929912/app/services/journals/create_service.rb#L883-L890) method, so that journals are not aggregated when a field is activated (I think deactivation can be aggregated).

We do a similar journal split in case there are multiple notes ("comments") on the same journal. The aggregation is skipped when the user makes multiple comments within the aggregation window.

On a side note, I would rather have the PR unrelated changes in separate PR's in the future, because it makes reviewing a lot easier.

Overall it's a terrific job, thanks @toy 👏

toy committed:
reorder results keys (start_date, end_date) for easier reading

toy committed:
put association class names in a constant

toy commented:
@dombesz Thank you for thorough review

toy committed:
use touch_and_save_journals

toy committed:
remove grouping option for multiple_attributes_changes

toy committed:
add spec for multiple_attributes_changes with multiple_values: true for consistency

toy committed:
delegate to CreateService instead of extracting methods to Helpers module

toy committed:
don't rescue in event_params

previously it would anyway fail in calling method, as nil was returned
instead of empty Hash

toy committed:
move falsey check of fetcher scope

dombesz reviewed (approved):
Works nicely, thanks @toy !
